### PR TITLE
Introduce parsec_taskpool_wait and parsec_taskpool_test

### PR DESCRIPTION
**Pull request by devreal**
From the TTG branches...

@therault please review carefully, there were some conflicts that I had to resolve...

Signed-off-by: Thomas Herault <herault@icl.utk.edu>
Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

### CODE REVIEW

**Comment by devreal** [comment]
Ahh, this seems to depend on https://github.com/ICLDisco/parsec/pull/391

**Comment by therault** [comment]
We should review this PR once the termination detection PR is merged.

File: parsec/runtime.h
Comment (abouteiller) [inline parsec/runtime.h:352]: Cannot the test set the task pool as completed (i.e., have the effect of an implicit wait), from that documentation I infer that the only correct code is to follow `taskpool_test` with `taskpool_wait` even when `taskpool_test` returned 0.

File: parsec/scheduling.c
Comment (abouteiller) [inline parsec/scheduling.c:601]: I do not understand why when we work on a specific task pool, we detach all task pools. Should we just call `parsec_dtd_detach_taskpool_from_context(tp)` here?

File: parsec/scheduling.c
Comment (abouteiller) [inline parsec/scheduling.c:946]: This is the only call site, and the `es` argument is set to a "constant" that can be deduced from `tp->context`; is the second argument to `__parsec_taskpool_test` needed?

File: tests/taskpool_wait/CMakeLists.txt
Comment (abouteiller) [inline tests/taskpool_wait/CMakeLists.txt:1]: All these tests should be in the directory `tests/api/taskpool_wait/`

File: tests/CMakeLists.txt
Comment (abouteiller) [inline tests/CMakeLists.txt:78]: This add_subdirectory must be moved inside the `TARGET parsec-ptgpp` block above

**Review by abouteiller — CHANGES_REQUESTED** [review]

File: parsec/scheduling.c
Comment (therault) [reply parsec/scheduling.c:601]: I will take a DTD test that reuses a DTD taskpool in a loop, and change the parsec_context_wait to a parsec_taskpool_wait, and test.

It seems obvious that we need to detach the taskpool if it is a DTD taskpool, and not the other taskpools, but unclear what else we should do.

File: parsec/scheduling.c
Comment (therault) [reply parsec/scheduling.c:946]: Probably it's a place where we should use `parsec_my_execution_stream()`

Comment (bosilca) [reply parsec/scheduling.c:946]: done.

File: tests/CMakeLists.txt
Comment (bosilca) [reply tests/CMakeLists.txt:78]: only because the single test in the directory uses both PTG andDTD taskpools.

File: tests/taskpool_wait/CMakeLists.txt
Comment (bosilca) [reply tests/taskpool_wait/CMakeLists.txt:1]: done.

**Comment by abouteiller** [comment]
Should this be slated for v4.0 release?

File: parsec/interfaces/dtd/insert_function.c
Comment (abouteiller) [inline parsec/interfaces/dtd/insert_function.c:230]: Could use the macro LIST_ITEM_SINGLETON here

File: parsec/data_distribution.c
Comment (abouteiller) [inline parsec/data_distribution.c:82]: This is way too high a debug level for this, should use parsec_debug_verbose(...)

File: parsec/interfaces/dtd/insert_function_internal.h
Comment (abouteiller) [inline parsec/interfaces/dtd/insert_function_internal.h:231]: I don't understand this comment

File: parsec/runtime.h
Comment (abouteiller) [reply parsec/runtime.h:352]: is the following mockup correct and does that description prevents it? 

```
taskpool_test(tp)
context_wait(ctx)
```

File: parsec/scheduling.c
Comment (abouteiller) [inline parsec/scheduling.c:587]: don't think that assert is correct, what about recursive task pools

File: parsec/scheduling.c
Comment (abouteiller) [inline parsec/scheduling.c:540]: don't think that assert is correct, what about recursive task pools
(same comment appears twice)

File: parsec/scheduling.c
Comment (abouteiller) [inline parsec/scheduling.c:552]: Shouldn't this do the same as `test` below?
```
/* Detach the current taskpool from the associated context */
     parsec_dtd_dequeue_taskpool(tp);
     PARSEC_LIST_ITEM_SINGLETON(tp);
```

File: tests/api/taskpool_wait/CMakeLists.txt
Comment (abouteiller) [inline tests/api/taskpool_wait/CMakeLists.txt:8]: name of test target (2Dband) is wrong

**Review by abouteiller — CHANGES_REQUESTED** [review]
Mo' review mo problems

File: parsec/data_distribution.c
Comment (bosilca) [reply parsec/data_distribution.c:82]: This is a real issue, it might or might not work we have no way to know if there will be collision or not between different data collections. So I would rather have this well visible, such that an educated user can fix their code (once it gets annoyed by the output).

File: parsec/interfaces/dtd/insert_function.c
Comment (bosilca) [reply parsec/interfaces/dtd/insert_function.c:230]: At this point I'm not sure the macro remains necessary in the code base, it always defaults to calling the `parsec_list_item_singleton ` function.

File: parsec/interfaces/dtd/insert_function_internal.h
Comment (bosilca) [reply parsec/interfaces/dtd/insert_function_internal.h:231]: Reminder that we do not need to manipulate this field with atomic operations.

File: parsec/runtime.h
Comment (bosilca) [reply parsec/runtime.h:352]: This is the expected usage, as `taskpool_test` does not alter the state of the context (aka. it remains active).

File: parsec/scheduling.c
Comment (bosilca) [reply parsec/scheduling.c:540]: We could allow usage of this function in other threads than master. That being said I'm not sure all accessors used in this function are thread safe.

File: parsec/scheduling.c
Comment (bosilca) [reply parsec/scheduling.c:552]: With the new API we designed few days ago, the use of `parsec_detach_all_dtd_taskpool_from_context` will not be necessary.

File: parsec/scheduling.c
Comment (bosilca) [reply parsec/scheduling.c:587]: Unlike in test I don't think we should allow any other threads to do a blocking operation on any parsec_objects. We will run out of working threads, and encounter undesired side-effects.

**Comment by therault** [comment]
I'm fixing issues in this pull request. I have a question for all:

Some DSLs (DTD, but others probably will also need the feature) require that something special is done on their taskpools when the user thread enters `context_wait()` or `taskpool_wait()`, and something should also be done when leaving `context_wait()` or `taskpool_wait()`.

For `taskpool_wait()`, it's easy, because we have the taskpool, so we can add a DSL function in the taskpool that the DSL registers.

For `context_wait()`, it's not that easy, because the context does not have a list of taskpools.

The current solution (in master) is that we call some DSL-specific functions (namely `parsec_detach_all_dtd_taskpool_from_context`), and the DSL keeps a list of taskpools. 

I want to make this more generic. I see two ways:

  - the DSL registers a callback to be called in the context. If the callback is already registered, it's the responsibility of the DSL not to register the callback again, and the DSL keeps a private list of taskpools that needs to be managed, similar to what is done today (except that instead of calling `parsec_detach_all_dtd_taskpool_from_context()` directly, we call `dsl_callback_enter_wait()` for each registered DSL.
  - each taskpool has an optional `dsl_enter_wait_cb()` and `dsl_leave_wait_cb()` function, and the context keeps a list of attached taskpools, and iterates over that list every time it enters / leaves context_wait (and we can use the same callbacks for the `taskpool_wait` version).

I'm more in favor of the second approach, because it's less state to manage at the DSL level, but maybe you have a different opinion?

File: parsec/interfaces/dtd/insert_function.c
Comment (abouteiller) [reply parsec/interfaces/dtd/insert_function.c:230]: the macro does the list_item cast, that's all it does. If we think this is not useful, we should remove the macro then.

File: parsec/data_distribution.c
Comment (abouteiller) [reply parsec/data_distribution.c:82]: looking at the code, I think you are referring to the line 64: registering the same key twice (this is abnormal but harmless, `parsec_info` is adequate), on line 68, the `DEBUG_VERBOSE` should be promoted to a `parsec_warning` because this is a real coding error that will cause problems, this line is unconditionally printing the registration, this is too verbose.

File: parsec/interfaces/dtd/insert_function.c
Comment (bosilca) [reply parsec/interfaces/dtd/insert_function.c:230]: ok, let me keep the code consistent with the rest (I'll switch to PARSEC_LIST_ITEM_SINGLETON)

File: parsec/data_distribution.c
Comment (bosilca) [reply parsec/data_distribution.c:82]: ok, I will make the change. It is unfortunate that the warnings are only delivered when in heavy debug mode.

Comment (therault) [reply parsec/data_distribution.c]: See 52caa8bee925eac4c3876079444690662393f69a

File: parsec/interfaces/dtd/insert_function.c
Comment (therault) [reply parsec/interfaces/dtd/insert_function.c]: Already fixed in the current code

File: parsec/scheduling.c
Comment (therault) [reply parsec/scheduling.c:543]: I think adding the feature that this function can be called from other threads than the master is outside the scope of this PR... As noted by George, there are accesses that are unsafe in this function, and we should keep the assert for now.

File: parsec/interfaces/dtd/parsec_dtd_data_flush.c
Comment (devreal) [reply parsec/interfaces/dtd/parsec_dtd_data_flush.c:322]: Actually every process sees that task. It's not inserted on just one process but only if the last writer of the  tile is not the same as the owner.

Comment (bosilca) [reply parsec/interfaces/dtd/parsec_dtd_data_flush.c:322]: Yes, I misread the code and @therault pointed me on the right direction. Why is the sending task missing one ref count ?

Comment (devreal) [reply parsec/interfaces/dtd/parsec_dtd_data_flush.c:322]: So here is what I found while debugging:

The flush task is created (https://github.com/ICLDisco/parsec/blob/c4cc5b01e46552cef018f90f05bd601ec6d0b350/parsec/interfaces/dtd/parsec_dtd_data_flush.c#L266) "manually" (i.e., not using `parsec_dtd_insert_task`) and then passed to `parsec_insert_dtd_flush_task` where it is retained (https://github.com/ICLDisco/parsec/blob/master/parsec/interfaces/dtd/parsec_dtd_data_flush.c#L134) and released again (https://github.com/ICLDisco/parsec/blob/master/parsec/interfaces/dtd/parsec_dtd_data_flush.c#L239) at the end of the function. That's according to plan, as far as I can see. When the second task of the pair is created, that first task is retained when it's new descendant is set (https://github.com/ICLDisco/parsec/blob/master/parsec/interfaces/dtd/parsec_dtd_data_flush.c#L185). This is good, because eventually this reference will be released once the dependency is released. However, we also release that first task right away because it's not the last writer anymore (https://github.com/ICLDisco/parsec/blob/master/parsec/interfaces/dtd/parsec_dtd_data_flush.c#L225), which is correct (we don't care about tasks anymore if they are not the last writer anymore). It's important though to keep that one reference that will be released once the dependency is released. That's the reference that is added if `is_first_flush_task` is `true`.

Maybe the original sin is that the code doesn't use `parsec_dtd_insert_task` (which hopefully would do just the right thing) but I cannot properly judge that. And I don't know why it wasn't done that way in the first place. My sense is that my solution is a hack, piled on top of more hacks. The manual ref-counting (e.g., https://github.com/ICLDisco/parsec/blob/c4cc5b01e46552cef018f90f05bd601ec6d0b350/parsec/interfaces/dtd/parsec_dtd_data_flush.c#L275) makes reasoning about the life-time hard. I'd prefer doing this properly but I'm afraid that pulling one string will unravel all others.

File: parsec/interfaces/dtd/insert_function.c
Comment (therault) [inline parsec/interfaces/dtd/insert_function.c:2225]: This is suspicious and breaking profiling.

File: parsec/interfaces/dtd/insert_function.c
Comment (therault) [reply parsec/interfaces/dtd/insert_function.c:1816]: Sorry, used this while debugging, forgot to remove it. Done.

File: parsec/interfaces/dtd/insert_function.c
Comment (therault) [reply parsec/interfaces/dtd/insert_function.c]: Sorry, used it while debugging, forgot to remove it. Done

File: parsec/interfaces/dtd/parsec_dtd_data_flush.c
Comment (therault) [reply parsec/interfaces/dtd/parsec_dtd_data_flush.c:322]: I believe this conversation (tile owner rank != tile last user rank) is resolved. Re-open it if I'm wrong?

**Comment by therault** [comment]
The last bug report we have had on this PR was from Sameer, but his case re-uses the same flow as RW in multiple inputs of a task, and it seems the issue is also present on master. A workaround has been found (by passing the not tracking flag to additional copies of the flow), and this should be solved, but should be solved in a separate PR.

Please review the PR, so that we can merge it.

@abouteiller your last review asks for changes, I believe they are implemented, please re-review and change your status.

**Review by bosilca — APPROVED** [review]

**Pull request merged**
